python/src/uagents/envelope.py

```diff
@@ -110,3 +117,49 @@ def _digest(self) -> bytes:
         if self.nonce is not None:
             hasher.update(struct.pack(">Q", self.nonce))
         return hasher.digest()
+
+
+class EnvelopeHistoryEntry(BaseModel):
+    timestamp: int = Field(default_factory=lambda: int(time.time()))
+    version: int
+    sender: str
+    target: str
+    session: UUID4
+    schema_digest: str
+    protocol_digest: Optional[str] = None
+    payload: Optional[str] = None
+
+    @field_serializer("session")
+    def serialize_session(self, session: UUID4, _info):
+        return str(session)
+
+    @classmethod
+    def from_envelope(cls, envelope: Envelope):
+        return cls(
+            version=envelope.version,
+            sender=envelope.sender,
+            target=envelope.target,
+            session=envelope.session,
+            schema_digest=envelope.schema_digest,
+            protocol_digest=envelope.protocol_digest,
+            payload=envelope.decode_payload(),
+        )
+
+
+class EnvelopeHistory(BaseModel):
+    envelopes: List[EnvelopeHistoryEntry]
+
+    def add_entry(self, entry: EnvelopeHistoryEntry):
+        self.envelopes.append(entry)
+        self.apply_retention_policy()
+
+    def apply_retention_policy(self):
+        """Remove entries older than 24 hours"""
+        cutoff_time = time.time() - 86400
+        self.envelopes = [e for e in self.envelopes if e.timestamp > cutoff_time]
+
+    @field_serializer("envelopes", when_used="always")
+    def serialize_envelopes_in_order(
+        self, envelopes: List[EnvelopeHistoryEntry], _info
+    ):
+        return sorted(envelopes, key=lambda e: e.timestamp)
```

Comment:
I'm wondering if we could manage both the retention policy and the serialization somewhat more efficiently by storing the envelopes as a `dict` with timestamp as the key?

Reply:
Good point. I'd argue that it doesn't matter much if we use models instead of dictionaries but I see a problem with the loop in line 159 now since that can get inefficient with a large number of messages within the retention period.
Since we decided against a maximum number of messages it's hard to identify an upper ceiling so some form of iteration over the dataset will be necessary.

We'll need a Model for validation in the Rest endpoint anyway so it made sense to do the sorting implicitly here. (although in my tests sorting isn't actually needed as the messages will be added to the list in the order they appear...)

Reply:
Yes, that line 159 was mostly what I had in mind. With the envelopes sorted by timestamp, we could start from the oldest and stop at the first envelope whose timestamp puts it inside the retention window.

I think that line 165 could then become something like `return list(envelopes.values())`.

Reply:
refactored the retention mechanism and dropped the serialiser method as that was not needed any more.